File: compiler/queries/comments.go
package queries

import (
	"slices"
	"wipple/database"
	"wipple/nodes/statements"
	"wipple/nodes/types"
	"wipple/typecheck"
	"wipple/visit"
)

func ErrorInstance(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(bound typecheck.ResolvedBound, comments CommentsData, trace []typecheck.Constraint)) {
	bounds, ok := database.GetFact[typecheck.BoundsFact](node)
	if !ok {
		return
	}

	for _, bound := range bounds {
		if bound.Instance != nil && bound.Error {
			instance, ok := bound.Instance.(*statements.InstanceDefinitionNode)
			if !ok {
				continue
			}

			comments := CommentsData{
				Nodes:    []database.Node{instance},
				Comments: instance.Comments,
				Links:    getLinks(db, instance, node, filter),
			}

			var trace []typecheck.Constraint
			for _, ty := range *bound.Bound.Substitutions {
				if node, ok := ty.(database.Node); ok {
					if fact, ok := database.GetFact[typecheck.TypedFact](node); ok && fact.Group != nil {
						trace = append(trace, fact.Group.Trace...)
						comments.Nodes = append(comments.Nodes, fact.Group.Nodes...)
					}
				}
			}

			slices.SortStableFunc(trace, func(left typecheck.Constraint, right typecheck.Constraint) int {
				return database.CompareSpans(left.Info().Span, right.Info().Span)
			})

			f(bound.Bound, comments, trace)
		}
	}
}

type CommentsData struct {
	Nodes    []database.Node
	Comments []string
	Links    map[string]Link
}

func Comments(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(data CommentsData)) {
	if defined, ok := database.GetFact[visit.DefinedFact](node); ok {
		f(CommentsData{
			Nodes:    []database.Node{node},
			Comments: defined.Definition.GetComments(),
			Links:    getLinks(db, node, node, filter),
		})
	} else if resolved, ok := database.GetFact[visit.ResolvedFact](node); ok && len(resolved.Definitions) == 1 {
		definition := resolved.Definitions[0]

		f(CommentsData{
			Nodes:    []database.Node{definition.GetNode()},
			Comments: definition.GetComments(),
			Links:    getLinks(db, definition.GetNode(), node, filter),
		})
	}

}

type Link struct {
	Node    database.Node
	Related []database.Node
	Types   []*typecheck.ConstructedType
}

func getLinks(db *database.Db, node database.Node, source database.Node, filter func(node database.Node) bool) map[string]Link {
	links := map[string]Link{}

	fact, ok := database.GetFact[visit.TypeParametersFact](node)
	if !ok {
		return links
	}

	for _, typeParameter := range fact {
		typeParameter, ok := typeParameter.(*types.TypeParameterNode)
		if !ok {
			continue
		}

		instantiated, ok := database.ContainsFact(db, func(node database.Node, fact typecheck.InstantiatedFact) (database.Node, bool) {
			if instantiated, ok := database.GetFact[typecheck.InstantiatedFact](node); ok {
				if instantiated.From == typeParameter && instantiated.Source == source {
					return node, true
				}
			}

			return nil, false
		})

		if !ok {
			continue
		}

		fact, ok := database.GetFact[typecheck.TypedFact](instantiated)
		if !ok || fact.Group == nil {
			continue
		}

		var uses []database.Node
		for _, node := range fact.Group.Nodes {
			if filter(node) {
				if _, ok := database.GetFact[typecheck.TypedFact](node); ok {
					uses = append(uses, node)
				}
			}
		}

		if len(uses) == 0 {
			continue
		}

		// Prefer to use non-hidden nodes if possible
		if slices.ContainsFunc(uses, func(node database.Node) bool {
			return !database.IsHiddenNode(node)
		}) {
			uses = slices.DeleteFunc(uses, database.IsHiddenNode)
		}

		slices.SortStableFunc(uses, func(left database.Node, right database.Node) int {
			return OrderGroupNodes(left) - OrderGroupNodes(right)
		})

		database.SortByProximity(uses, source)

		links[typeParameter.Name] = Link{
			Node:    uses[0],
			Related: uses,
			Types:   fact.Group.Types,
		}
	}

	return links
}

File: compiler/src/queries/comments.rs
use crate::{
    database::NodeRef,
    nodes::{InstanceDefinitionNode, TypeParameterNode},
    queries::QueryCtx,
    typecheck::{Bound, Bounds, Constraint, ConstructedType, Instantiated, Type, Typed},
    visit::{Defined, Resolved, TypeParameters},
};
use std::collections::HashMap;

#[derive(Debug)]
pub struct QueriedErrorInstance {
    pub bound: Bound,
    pub comments: QueriedComments,
    pub trace: Vec<Box<dyn Constraint>>,
}

pub fn error_instance(ctx: &QueryCtx<'_>, f: &mut dyn FnMut(QueriedErrorInstance)) {
    let Some(Bounds(bounds)) = ctx.db.get(&ctx.node) else {
        return;
    };

    for item in bounds {
        if let Some(instance) = item.instance
            && instance.error
        {
            let Some(instance_definition) = instance.node.downcast_ref::<InstanceDefinitionNode>()
            else {
                continue;
            };

            let mut comments = QueriedComments {
                nodes: vec![instance.node.clone()],
                comments: instance_definition.comments.clone(),
                links: get_links(ctx, &instance.node),
            };

            let trace = item
                .bound
                .substitutions
                .values()
                .into_iter()
                .filter_map(|ty| {
                    let Type::Node(node) = ty else {
                        return None;
                    };

                    let Some(Typed { group: Some(group) }) = ctx.db.get(&node) else {
                        return None;
                    };

                    comments.nodes.extend(group.nodes.iter().cloned());

                    Some(group.trace.clone())
                })
                .flatten()
                .collect::<Vec<_>>();

            f(QueriedErrorInstance {
                bound: item.bound,
                comments,
                trace,
            });
        }
    }
}

#[derive(Debug)]
pub struct QueriedComments {
    pub nodes: Vec<NodeRef>,
    pub comments: Vec<String>,
    pub links: HashMap<String, QueriedCommentsLink>,
}

pub fn comments(ctx: &QueryCtx<'_>, f: &mut dyn FnMut(QueriedComments)) {
    let definition_node = ctx.db.get::<Resolved>(&ctx.node).map_or_else(
        || ctx.node.clone(),
        |Resolved { definitions, .. }| definitions.into_iter().next().unwrap(),
    );

    let Some(Defined(definition)) = ctx.db.get::<Defined>(&definition_node) else {
        return;
    };

    f(QueriedComments {
        nodes: vec![definition.node()],
        comments: definition.comments().to_vec(),
        links: get_links(ctx, &definition.node()),
    })
}

#[derive(Debug)]
pub struct QueriedCommentsLink {
    pub node: NodeRef,
    pub related: Vec<NodeRef>,
    pub types: Vec<ConstructedType>,
}

fn get_links(
    ctx: &QueryCtx<'_>,
    definition_node: &NodeRef,
) -> HashMap<String, QueriedCommentsLink> {
    let mut links = HashMap::new();

    let Some(TypeParameters(type_parameters)) = ctx.db.get(definition_node) else {
        return links;
    };

    for type_parameter_node in type_parameters {
        let Some(type_parameter) = type_parameter_node.downcast_ref::<TypeParameterNode>() else {
            continue;
        };

        let Some(instantiated) = ctx
            .db
            .iter::<Instantiated>()
            .find_map(|(node, instantiated)| {
                (instantiated.from == type_parameter_node && instantiated.source_node == ctx.node)
                    .then_some(node)
            })
        else {
            continue;
        };

        let Some(Typed { group: Some(group) }) = ctx.db.get(&instantiated) else {
            continue;
        };

        let mut uses = group.nodes.iter().cloned();
        let first = uses.next().unwrap();
        let related = uses.collect::<Vec<_>>();

        links.insert(
            type_parameter.name.clone(),
            QueriedCommentsLink {
                node: first,
                related,
                types: group.types.clone(),
            },
        );
    }

    links
}
